eclair-core/src/main/scala/fr/acinq/eclair/payment/send/Recipient.scala

```diff
@@ -0,0 +1,185 @@
+
+  override def buildPayloads(paymentHash: ByteVector32, route: Route): Try[PaymentPayloads] = {
+    val finalPayload = NodePayload(nodeId, FinalPayload.Standard.createKeySendPayload(route.amount, totalAmount, expiry, preimage, customTlvs))
+    Success(OutgoingPaymentPacket.buildPayloads(totalAmount, expiry, finalPayload, route.hops))
+  }
+}
+
+/**
+ * A payment recipient that can be reached through a trampoline node (such recipients usually cannot be found in the
+ * public graph). When a payment needs to be split across multiple trampoline nodes, the sender must create multiple
+ * recipient objects with distinct partial amounts, trampoline payment secrets and trampoline hops.
+ *
+ * Note that we don't need to support the case where we'd use multiple trampoline hops in the same route: since we have
+ * access to the network graph, it's always more efficient to find a channel route to the last trampoline node.
+ */
+case class ClearTrampolineRecipient(invoice: Bolt11Invoice,
+                                    partialAmount: MilliSatoshi,
+                                    totalAmount: MilliSatoshi,
+                                    expiry: CltvExpiry,
+                                    trampolineHop: NodeHop,
+                                    trampolinePaymentSecret: ByteVector32,
+                                    customTlvs: Seq[GenericTlv] = Nil) extends Recipient {
+  require(trampolineHop.nextNodeId == invoice.nodeId, "trampoline route must end at the recipient")
+
+  override val nodeId = invoice.nodeId
+  override val features = invoice.features
+  override val extraEdges = Nil
+
+  // We cannot simply inject virtual edges in the graph and let path-finding dynamically select when to use trampoline
+  // hops because that would break MPP aggregation at the trampoline node: the fees of the trampoline hop should be paid
+  // only once, even if we need to split our payment towards the trampoline node.
+  //
+  //   +-----> Bob ------+
+  //   |                 |
+  // Alice             Teddy -----> Dave
+  //   |                 |
+  //   +-----> Carol ----+
+  //
+  // It's thus simpler to decide beforehand how to split a payment across different trampoline hops, and then handle the
+  // payments to the distinct trampoline nodes separately: from the sender's point of view, those payments behave like
+  // standard multi-part payments where the trampoline node acts as the final recipient.
+  private val trampolineNodeId = trampolineHop.nodeId
+  private val trampolineAmount = partialAmount + trampolineHop.fee(partialAmount)
+  private val trampolineExpiry = expiry + trampolineHop.cltvExpiryDelta
+  override val targetNodeId = trampolineNodeId
+  override val targetTotalAmount = trampolineAmount
+
+  override def targetMaxFee(routeParams: RouteParams): MilliSatoshi = routeParams.getMaxFee(partialAmount) - trampolineHop.fee(partialAmount)
```

Comment:
I don't think this is right, on the contrary. This ensures that the trampoline fee is deduced from the fee budget we allocate to the MPP payment sent _to_ the trampoline node. This way the complete fee budget fits into what is allocated in `routeParams`. Even if we split across two trampoline nodes, this would work, because we would use the same trick as what we do in `MultiPartPaymentLifecycle`, where we update `routeParams` to deduce the fees already sent to pending routes?

Reply:
Just to make sure I understand your example, you are talking about splitting across two different trampoline nodes (which isn't supported yet), right? In that case we would split the fee budget between those two, and tell the first one that it has a fee budget of 5 sats and the other ones has a fee budget of 5 sat, similar to what we do for MPP here: https://github.com/ACINQ/eclair/blob/3b2eb964fd51557cf59e39c897f7890ce55dadaa/eclair-core/src/main/scala/fr/acinq/eclair/payment/send/MultiPartPaymentLifecycle.scala#L424

What this line of code guarantees if that if the first trampoline hop costs 1 sat, we will only use up to 4 sats when finding a set of MPP routes to that trampoline hop. If the second trampoline hop costs 2 sats, we will only use up to 3 sats when finding a set of MPP routes to that trampoline hop. Does that make sense?

Reply:
OK, that works. But that would require to be careful when we process each trampoline part to do it sequentially.
I assumed that we would create both recipients and then route to these recipients, instead we need to create the first recipient, route to it, compute the remaining fee budget and only then create the second recipient. If we're clear about that then it's fine.

Reply:
We can do it sequentially, or just split it equally among trampoline nodes and give them the same ratio of the total budget. I don't know yet what will be best here, but we'll probably figure it out when we do that multi-trampoline support (and if we need to refactor this `targetMaxFee` function it's ok, it's not a fundamental part of the design of `Recipient`.